### PR TITLE
new blog post: Bullets are good for your coq proofs

### PR DESCRIPTION
**Pull request by gasche**
cc @dbp @bennn @maxsnew

### CODE REVIEW

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+proof of first subgoal.
+
+proof of second subgoal.
+```
+
+There are many ways to structure this to make the structure more
+apparent: people would typically have a comment on each subgoal, or
+make disciplined use of indentation and blank lines. But, in my
+experience, a major problem with this style was maintenability in the
```
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: `maintenability` -> `maintainability`

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+face of changes to the definitions or parts of automation. It could be
+very hard of what was happening when a proof suddenly broke after
```
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: ❗️  I don't understand this line. Is the purpose of this sentence just to say "For example:" ?

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  then apply to the beginning of the second subgoal, and you get very
+  confusing errors again.
+
+What we need for robustness is a way to indicate our *intent* to Coq,
+when we think that a subgoal is finished and that a new subgoal
+starts, so that Coq can fail loudly at the moment where it notices
+that this intent does not match reality, instead of at an arbitrary
+later time.
+
+(The `S*Case` tactics used in Software Foundations can solve this
```
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: SF: https://www.cis.upenn.edu/~bcpierce/sf/

I tried to find a link to S*Case from SF, but it seems they recommend bullets now. Yay!
https://www.cis.upenn.edu/~bcpierce/sf/current/Basics.html

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+```
+{ induction foo. (* this creates many subgoal *)
+  {
+    proof of first subgoal.
+  }
+  {
+    proof of second subgoal.
+  }
+} (* would fail if a new subgoal appeared *)
```
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: "would fail if there were more than two subgoals"

(The other subgoal isn't appearing, it's just from `induction foo`)

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+```
+
+## Bullets
```
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: Maybe, "List Bullets" ?

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+- proof of second remaining subgoal.
+  rest of the proof of the second subgoal.
+}
+```
+
+(Notice that the subgoal-creating step is vertically aligned with the
+proof steps: I use both braces and bullets, but take only one
+indentation level each time.)
+
+As an exception, I may omit the braces around if we are at the
```
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: ~~around~~

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+}
+```
+
+(This is the only time where I use several bullet levels.)
+
+If you are the kind of programmer that is passionate about indentation
+style, I should now have tricked you to use bullets to propose
+a different variant. Otherwise, please consider using bullet anyway,
+for example by following the style above, it will make your life
+easier in the face of changing developments.
```
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: Hmmm, want to encourage people to email you success stories? (Either here, or on Reddit etc. ?)

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md (hunk 4 above)
Comment (gasche) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: The idea is that you may add a constructor to the inductive, which would create a new subgoal in each induction. I will try to clarify that before.

Comment (gasche) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: > If you also want to make sure that an error occurs if the number of
subgoals changes (for example if new constructors are added to the
inductive type of `foo`), you can use an outer layer of braces:

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md (hunk 7 above)
Comment (gasche) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: Meh. I think if we don't have comments it does not make much sense to invite comments. (Nothing against reddit of course, although I wouldn't post there myself as I avoid self-promotion.)

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md (hunk 4 above)
Comment (bennn) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: yes, cool

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+cases and subgoals, and then I refine them to add more automation for
+robustness. I found bullets very useful for the first step, and during
+the refinement process.
+
+# Bullets
+
+Bullets are actually a combination of two features, braces `{ ... }`
+and actually list bullets, `-`, `+`, or `*` or homogeneous repetitions
+of those, for example `--` or `***`.
+
```
Comment (maxsnew) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: This sentence is confusing to me. It says there's 2 features but seemingly mentions 3, plus the "or `*` or" makes it hard to parse. Maybe you meant the word "actual" instead of "actually"?

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+This gives a large space of freedom in how you want to use these
+features. You can use only braces, only bullets, braces and only one
+level of bullets, etc. My own style evolved with experience using the
+feature, and I will present the current status below.
+
+# My current bullet style
+
+When deciding how to use bullets, one distinguishes the commands that
+preserve the number of subgoals and those that may create new
+subgoals. I do some further distinctions.
```
Comment (maxsnew) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: "I use some additional distinctions" or "I employ some additional distinctions" would be better here

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+  finish the sub-subgoal.
+- third subgoal.
+}
+```
+
+(This is the only time where I use several bullet levels.)
+
+If you are the kind of programmer that is passionate about indentation
+style, I should now have tricked you to use bullets to propose
+a different variant. Otherwise, please consider using bullet anyway,
```
Comment (maxsnew) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: bullets should be plural here.

File: blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
++ remaining sub-subgoal of the second subgoal.
+  finish the sub-subgoal.
+- third subgoal.
+}
+```
+
+(This is the only time where I use several bullet levels.)
+
+If you are the kind of programmer that is passionate about indentation
+style, I should now have tricked you to use bullets to propose
```
Comment (maxsnew) [inline blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md]: Don't really follow this sentence, why would someone interested in indentation style want to use bullets?

**Comment by gasche** [comment]
@maxsnew: thanks, I updated the post.
